src/lake/Lake/Load/Main.lean

```diff
@@ -148,33 +148,44 @@ downloading and/or updating them as necessary.
     logWarning <|
-      "manifest out of date: packages directory changed, " ++
-      "use `lake update` to update"
+      "manifest out of date: packages directory changed; " ++
+      "use `lake update` to rebuild the manifest (warning: this will update ALL workspace dependencies)"
   let relPkgsDir := manifest.packagesDir?.getD ws.relPkgsDir
+  let pkgEntries := manifest.packages.foldl (init := mkNameMap PackageEntry)
+    fun map entry => map.insert entry.name entry
   let res ← EStateT.run' (mkNameMap Package) do
     buildAcyclic (·.name) ws.root fun pkg resolve => do
       let topLevel := pkg.name = ws.root.name
       let deps ← IO.ofExcept <| loadDepsFromEnv pkg.configEnv pkg.leanOpts
       if topLevel then
+        if manifest.isEmpty && !deps.isEmpty then
+          error "missing manifest; use `lake update` to generate one"
         for dep in deps do
           let warnOutOfDate (what : String) :=
             logWarning <|
-              s!"manifest out of date: {what} of dependency '{dep.name}' changed, " ++
-              "use `lake update` to update"
-          if let .some entry := manifest.find? dep.name then
+              s!"manifest out of date: {what} of dependency '{dep.name}' changed; " ++
+              s!"use `lake update {dep.name}` to update it"
+          if let .some entry := pkgEntries.find? dep.name then
           match dep.src, entry with
           | .git (url := url) (rev := rev) .., .git (url := url') (inputRev? := rev')  .. =>
             if url ≠ url' then warnOutOfDate "git url"
             if rev ≠ rev' then warnOutOfDate "git revision"
           | .path .., .path .. => pure ()
           | _, _ => warnOutOfDate "source kind (git/path)"
       let depPkgs ← deps.mapM fun dep => fetchOrCreate dep.name do
-        let .some entry := manifest.find? dep.name
-          | error <| s!"dependency '{dep.name}' of '{pkg.name}' not in manifest, " ++
-            "use `lake update` to update"
-        let result ← entry.materialize ws.dir relPkgsDir
-        loadDepPackage ws.dir result pkg.leanOpts dep.opts reconfigure
-      return { pkg with opaqueDeps := ← depPkgs.mapM (.mk <$> resolve ·) }
+        if let some entry := pkgEntries.find? dep.name then
+          let result ← entry.materialize ws.dir relPkgsDir
+          loadDepPackage ws.dir result pkg.leanOpts dep.opts reconfigure
+        else if topLevel then
+          error <|
+            s!"dependency '{dep.name}' not in manifest; " ++
+            s!"use `lake update {dep.name}` to add it"
+        else
+          error <|
+            s!"dependency '{dep.name}' of '{pkg.name}' not in manifest; " ++
+            s!"check with the developer of {pkg.name} to see if they forgot to update their manifest file," ++
+            s!"or use `lake update` to generate a new, complete file (warning: this will update ALL workspace dependencies)"
+      return {pkg with opaqueDeps := ← depPkgs.mapM (.mk <$> resolve ·)}
```

Comment:
@semorrison I am curious what you think of my changes of the manifest-related error messages.